src/form.js

```diff
@@ -1172,6 +1184,80 @@ define([
         },
     };
 
+    function makeHashtagInferencer(form) {
+        function decorator(obj) {
+            var realHashtagToXPath = obj.hashtagToXPath;
+            obj.hashtagToXPath = function (hashtag) {
+                var xpath = realHashtagToXPath(hashtag);
+                if (xpath !== info.hashtagMap[hashtag]) {
+                    info.hashtagMap[hashtag] = xpath;
+                    changes[hashtag] = xpath;
+                }
+                return xpath;
+            };
+            return obj;
+        }
+
+        function makeTransform(hashtagPrefix, parentPath) {
+            function transform(prop) { return parentPath + "/" + prop; }
+            return [hashtagPrefix, function (prop) {
+                usedTransforms[hashtagPrefix] = transform;
+                return transform(prop);
+            }];
+        }
+
+        // Default case hashtag prefixes used to guess hashtag meanings.
+        var session = "instance('commcaresession')/session/data/case_id",
+            casepath = "instance('casedb')/cases/case[@case_id = " + session + "]",
+            parent = "instance('casedb')/cases/case[@case_id = " + casepath + "/index/parent]",
+            grandma = "instance('casedb')/cases/case[@case_id = " + parent + "/index/parent]",
+            user = "instance('casedb')/cases/case[@case_type = 'commcare-user']" +
+                   "[hq_user_id = instance('commcaresession')/session/context/userid]",
+            info = {
+                hashtagTransformations: _.object([
+                    makeTransform("#case/", casepath),
+                    makeTransform("#case/parent/", parent),
+                    makeTransform("#case/grandparent/", grandma),
+                    makeTransform("#user/", user),
+                ]),
+            },
+            changes = null, // reset on inferHashtags
+            usedTransforms = null, // reset on inferHashtags
+            parser = xpath.parser(info, decorator);
+
+        /**
+         * Infer hashtags using default hashtag transformations
+         *
+         * This function tries to make up for the fact that there was a
+         * time when xpath transform details were not written in the
+         * <vellum:hashtags> element of form XML, and therefore it is
+         * necessary to guess hashtag meansings using default transformations.
+         */
+        return function inferHashtags(hashtag, xpath_) {
+            var newpath;
+            changes = {};
+            usedTransforms = {};
+            info.hashtagMap = _.clone(form.hashtagMap);
+            info.hashtagNamespaces = form.hashtagNamespaces;
+            try {
+                newpath = parser.parse(hashtag).toXPath();
```

Comment:
This is where the magic happens. The hashtag is parsed and converted to xpath and then (below) compared to the original xpath expression. If they match, use the hashtag information gathered during the parse to update the form's hashtags and transformations.